utils/src/main/java/io/anlessini/utils/ImportCollection.java

```diff
@@ -0,0 +1,381 @@
+            input.getFileName().toString() + ": " + cnt + " docs added.");
+      } catch (Exception e) {
+        LOG.error(Thread.currentThread().getName() + ": Unexpected exception:", e);
+      } finally {
+        if (fileSegment != null) {
+          fileSegment.close();
+        }
+      }
+    }
+
+    private void sendBatchRequest() throws Exception {
+      int retries = 0;
+      TableWriteItems items = new TableWriteItems(args.dynamoTable)
+          .withItemsToPut(batch.stream().map(this::toDynamoDBItem).collect(Collectors.toList()));
+      LOG.info("Writing " + items.getItemsToPut().size() + " items to DynamoDB");
+
+      AmazonDynamoDB dynamoDBClient = dynamoPool.borrowObject();
+      DynamoDB dynamoDB = new DynamoDB(dynamoDBClient);
+      try {
+        BatchWriteItemOutcome outcome = null;
+        while (outcome == null) {
+          try {
+            outcome = dynamoDB.batchWriteItem(items);
+          } catch (ProvisionedThroughputExceededException e) {
+            LOG.error("Error sending BatchWriteItem request to DynamoDB", e);
+            if (retries++ < MAX_REQUEST_RETRIES) {
+              Thread.sleep((long) (Math.pow(2, retries) * 1000));
+            } else {
+              throw new RuntimeException("BatchWriteItem failed after too many retries", e);
+            }
+          } catch (ResourceNotFoundException | RequestLimitExceededException |
+              ItemCollectionSizeLimitExceededException | InternalServerErrorException e) {
+            LOG.error("Encountered non-retriable error", e);
+            throw new RuntimeException("BatchWriteItem failed due to non-retriable error", e);
+          }
+        }
+
+        Map<String, List<WriteRequest>> unprocessedItems = outcome.getUnprocessedItems();
+        retries = 0;
+        BatchWriteItemResult writeUnprocessedResult;
+        while (!unprocessedItems.isEmpty()) {
+          try {
+            writeUnprocessedResult = dynamoDBClient.batchWriteItem(unprocessedItems);
+            unprocessedItems = writeUnprocessedResult.getUnprocessedItems();
+          } catch (ProvisionedThroughputExceededException e) {
+            LOG.error("Error sending BatchWriteItem request to DynamoDB", e);
+            if (retries++ < MAX_REQUEST_RETRIES) {
+              Thread.sleep((long) (Math.pow(2, retries) * 1000));
+            } else {
+              throw new RuntimeException("BatchWriteItem failed after too many retries", e);
+            }
+          } catch (ResourceNotFoundException | RequestLimitExceededException |
+              ItemCollectionSizeLimitExceededException | InternalServerErrorException e) {
+            LOG.error("Encountered non-retriable error", e);
+            throw new RuntimeException("BatchWriteItem failed due to non-retriable error", e);
+          }
+        }
+      } finally {
+        dynamoPool.returnObject(dynamoDBClient);
+      }
+    }
+
+    public Item toDynamoDBItem(Document doc) {
```

Comment:
@shaneding I hope to get some of your inputs on this. This is the function that turns a Lucene `Document` to a DynamoDB `Item`. 

Both Lucene and DynamoDB are "schemaless". In Lucene, a Document have multiple fields of the same name, and these fields can all be of different types. And Documents need not have the same fields at all. Similar situation goes for DynamoDB. From what I have observed each record can be roughly taken as a key'ed JSON object.

Here in this function, I just take all the fields in a Lucene Document, group them by name, and write the values to a Dynamo Item in the form of String List (see Dynamo [data types](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes)). With the exception of three special fields, which I write as a String rather than a List: `id`, `contents`, and `raw` (because I know for sure that they appear only once). This implementation is different from [insert_acl_batch.py](https://github.com/castorini/anlessini/blob/master/pythonscripts/insert_acl_batch.py#L30), which seems to have knowledge about the fields' occurrences. 

Consequentially, any fields you request from DynamoDB would now be a List, even though you might know that there would be at most one element. My rationale is that the client usually has more knowledge about some specific fields than this dumb importer, and therefore it is easier to make assertions about them on the client side.

What do you think?

Reply:
One potential issue with this is what if the field itself is a list, for example `authors` for the ACL dataset will be a list already. Or perhaps am I missing something for the String List datatype?

Reply:
Why not create columns for the fields in dynamoDB, and then just extract items for those fields from the lucene document, then do insertion to dynamoDB?

Reply:
The Lucene Document would not have a field type that is already a List type. When you call `getFields(name)` on a Lucene Document, it iterates through its internal list of fields and find all fields with a matching name, and return it as an array/list.

In the case of `authors`, it is not a single field called `authors` with a List value type, but rather multiple fields called `authors` with a String value type.

Reply:
The method of

1. Creating a schema in Dynamo and then,
2. Query Lucene Document by the schema to perform insertion

will also requires us to explicitly define the schema in Dynamo for the collection supported. I mean there are quite a handful of collections and it would be non-trivial amount of work.